pkg/build/strategies/dockerfile/dockerfile.go

```diff
@@ -0,0 +1,323 @@
+			buffer.WriteString(fmt.Sprintf(" %s=%s", k, v))
+		}
+		buffer.WriteString("\n")
+	}
+
+	// only COPY scripts dir if required scripts are present, i.e. the dir is not empty;
+	// even if the "scripts" dir exists, the COPY would fail if it was empty
+	scriptsProvided, fileNames := checkValidDirWithContents(filepath.Join(builder.config.WorkingDir, builder.uploadScriptsDir))
+	assembleProvided := false
+	runProvided := false
+	for _, f := range fileNames {
+		glog.V(2).Infof("found override script file %s", f.Name())
+		if f.Name() == "run" {
+			runProvided = true
+		} else if f.Name() == "assemble" {
+			assembleProvided = true
+		}
+		if runProvided && assembleProvided {
+			break
+		}
+	}
+
+	if scriptsProvided {
+		glog.V(2).Infof("Override scripts are included in directory %q", builder.uploadScriptsDir)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadScriptsDir), sanitize(filepath.ToSlash(scriptsDestDir))))
+	} else {
+		// if an err on reading or opening dir, can't copy it
+		glog.V(2).Infof("Could not gather scripts from the directory %q", builder.uploadScriptsDir)
+	}
+
+	env := sti.CreateBuildEnvironment(builder.config.WorkingDir, config.Environment)
+	if len(env) > 0 {
+		buffer.WriteString("ENV")
+		for _, v := range env {
+			buffer.WriteString(fmt.Sprintf(" %s", sanitize(v)))
+		}
+		buffer.WriteString("\n")
+	}
+
+	// copy in the user's source code.
+	buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(builder.uploadSrcDir), sanitize(filepath.ToSlash(sourcesDestDir))))
+
+	glog.V(2).Infof("Processing secret inputs: %#v", config.Injections)
+	config.Injections = util.FixInjectionsWithRelativePath(config.ImageWorkDir, config.Injections)
+	glog.V(2).Infof("Processed secret inputs: %#v", config.Injections)
+
+	for _, injection := range config.Injections {
+		src := filepath.Join(api.Injections, injection.Source)
+		buffer.WriteString(fmt.Sprintf("COPY --chown=%s:0 %s %s\n", sanitize(imageUser), sanitize(src), sanitize(injection.Destination)))
+	}
+
+	if assembleProvided {
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(scriptsDestDir, "assemble"))))
+	} else {
+		buffer.WriteString(fmt.Sprintf("# Assemble script sourced from builder image based on user input or image metadata.\n"))
+		buffer.WriteString(fmt.Sprintf("# If this file does not exist in the image, the build will fail.\n"))
+		buffer.WriteString(fmt.Sprintf("RUN %s\n", sanitize(filepath.Join(config.ImageScriptsDir, "assemble"))))
+	}
+
+	// Remove any secrets that were copied into the image,
+	// but leave configmap content alone.
+	wroteRun := false
+	for _, injection := range config.Injections {
+		if injection.Keep {
+			continue
+		}
+		if !wroteRun {
+			buffer.WriteString("RUN ")
+			wroteRun = true
+		} else {
+			buffer.WriteString(" && ")
+		}
+		buffer.WriteString(fmt.Sprintf("rm -rf %s", sanitize(injection.Destination)))
```

Comment:
Noting change in behavior here, where secrets are deleted rather than truncated.
It's hard for me to see a use case where a developer relies on a truncated secret file existing, but this xkcd always comes to mind: https://xkcd.com/1172/

Reply:
yeah it's a fair point.  my guess is we'll phase in this new build approach over time so people will be able to fallback to the existing behavior if they want it, at least for some period of time.